src/types.md

```diff
@@ -236,3 +236,311 @@ fn main() {
+informs you that you have a problem.)
+
+> I never box things unless it’s really big - MG
+
+Another heuristic is if part of your data structure is very rarely filled,
+in which case you may wish to `Box` it to avoid incurring an overhead for all
+other instances of the type.
+
+```rust
+# struct Humility; struct Talent; struct Ego;
+struct Popstar {
+  ego: Ego,
+  talent: Talent,
+  humility: Option<Box<Humility>>,
+}
+# fn main() {}
+```
+
+(This is one reason why people like using [anyhow](https://docs.rs/anyhow/latest/anyhow/)
+for their errors; it means the failure case in their `Result` enum is only
+a pointer wide.)
+
+Of course, Rust may require you to use a box:
+
+* if you need to `Pin` some data, typically for async Rust
+* if you otherwise have an infinitely sized data structure.
+
+but as usual, the compiler will explain very nicely.
+
+## I miss operator overloading! What do I do?
+
+Implement the standard traits instead. This has equivalent effect in that
+folks will be able to use your type in a standard Rusty way without knowing
+too much special about your type.
+
+## Should I have public fields or accessor methods?
+
+The trade-offs are similar to C++ except that Rust's pattern-matching makes it
+very convenient to match on fields, so within a realm of code that you own you
+may bias towards having more public fields than you're used to. As with C++,
+this can give you a future compatibility burden.
+
+## When should I use a newtype wrapper?
+
+The [newtype wrapper pattern](https://rust-unofficial.github.io/patterns/patterns/behavioural/newtype.html)
+uses Rust's type systems to enforce extra behavior without necessarily changing
+the underlying representation.
+
+```rust
+# fn get_rocket_length() -> Inches { Inches(7) }
+struct Inches(u32);
+struct Centimeters(u32);
+
+fn build_mars_orbiter() {
+  let rocket_length: Inches = get_rocket_length();
+  // mate_to_orbiter(rocket_length); // does not compile because this takes cm
+}
+```
+
+Other examples that have been used:
+* An IP address which is guaranteed not to be localhost;
+* Non-zero numbers;
+* IDs which are guaranteed to be unique
+
+Such new types typically need a lot of boilerplate, especially to implement
+the traits which users of your type would expect to find. On the other hand,
+they allow you to use Rust's type system to statically prevent logic bugs.
+
+A heuristic: if there are some invariants you'd be checking for at runtime,
+see if you can use a newtype wrapper to do it statically instead. Although it
+may be more code to start with, you'll save the effort of finding and fixing
+logic bugs later.
+
+## How else can I use Rust's type system to avoid high-level logic bugs?
+
+Yes.
+
+*Zero sized types* ("ZSTs"). Often used as capability tokens - you can statically
+prove that code exclusively has the right to do something. For example:
+
+```rust
+struct PermissionToUseCamera;
+
+fn prompt_user_for_permission() -> Option<PermissionToUseCamera> {
```

Comment:
I'm a bit hesitant about this example, because while it shows the value of a ZST, in real life I would:

```
fn request_camera() -> Option<CameraDeviceProxy> {
  // ...
}

impl CameraDeviceProxy {
  pub fn take_picture(self, ...) {
    // ...
  }
}
```

Not sure if that matters.